### ISSUE TITLE
Need an Ivy discussion forum to ask user questions - e.g. where Alpha and Omega commands?

### ISSUE THREAD

**Issue by benjamin-rood** [robpike/ivy#28]
Hi Rob, sorry to bother you here, but I don't see how in Ivy you can use the APL α and ω commands, e.g. if I wanted to create a gcd operator like so:

![screen shot 2016-06-17 at 3 08 20 pm](https://cloud.githubusercontent.com/assets/10102132/16139651/d8f3713a-349d-11e6-80bf-f32f44a99e63.png)

src: http://dfns.dyalog.com/n_gcd.htm

**Comment by robpike** [comment]
You can't, at least not yet. Ivy is far from a complete mapping of APL's features, even APL\360. I believe α and ω came much later.

I may do something about that one day but have no plans to do so.

**Comment by benjamin-rood** [comment]
Thanks for the response. I wonder how one could do a recursive gcd operator then without conditionals? Ultimately I wanted to try to implement a binary coprime operator where `op x coprime n` produces a vector with all natural numbers ≤ n which are coprime to x.

I do appreciate that as you say, this is a plaything. I very much enjoyed the implementation in Go presentation and I learnt a lot from it, so thank you.

**Comment by robpike** [comment]
Could do that. It hasn't really had enough users until you started posting all these power-user videos :)

-rob

**Comment by fumin** [comment]
Could we not panic in [`value.Errorf`](https://github.com/robpike/ivy/blob/v0.2.7/value/value.go#L54)?
In particular, [`value.index`](https://github.com/robpike/ivy/blob/v0.2.7/value/index.go#L47) already does bounds checking, so repanicking seems wierd.

By the way, sincere gratitude for creating this wonderful piece of software!!
<details>
  <summary>A bit of background why I prefer not panicking</summary>
  
  I actually want to use ivy as a script language embedded in a Go programme.
  The Go programme is doing image manipulation on [Gerber](https://en.wikipedia.org/wiki/Gerber_format) images.
  In particular, it is trying to match parts of an image to a golden image.
  Due to the fragmentation of the PCB industry, there are myriad ways how our suppliers
  describe their manufactured parts.
  I can't handle all these different ways, so I am thinking of letting operators figure out the image mapping with their own eyeballs, and script it out to the programme.
  You may ask why not simply let operators input three numbers denoting the X, Y translations and theta for rotation.
  This is because in order to increase mechanical efficiency, some parts may contain multiple heads and work on many units simulatenously.
  In other words, the image mapping is something like:
  ```
  op panel p = 
    p[0] > 0 && p[1] > 0: p[0]*cos(theta) + p[1]*sin(theta)
    p[0] < 0 && p[1] > 0: ...
    ... 
  ```

  I was initially thinking of using https://github.com/pkujhd/goloader to allow operators to write in Go, but hacking Go's linker seems to fragile.
  Then I thought of `GOARCH=wasm`, but moving slices in and out of wasm is a pain in the ass.
  Moreover, `GOARCH=wasm` in it's current state is not much faster than an interpreter.
  In the end, I found ivy, which seems suited for these kinds of mathematical tasks.
  However, if ivy's `value.Errorf` panics, I would need to `recover` which feels like I am commiting a sin.
</details>

**Comment by fumin** [comment]
Why does `(5 5 rho iota 25)[3 2; 1 2 3]` return the following?

```
11 12
13  6
 7  8
```

I was expecting

```
11 12 13 
6 7  8
```

<details>
  <summary>Below is my investigation</summary>  
```
(5 5 rho iota 25)[3 2; 1]
11 6

(5 5 rho iota 25)[3 2; 1 2]
11 12
 6  7

(5 5 rho iota 25)[3 2; 1 2 3]
11 12
13  6
 7  8
```
</details>

**Comment by robpike** [comment]
Regarding the indexing example, that's the usual APL way, as you can see by trying the example on https://tryapl.org/. The vector left of the semicolon selects the first axis, while the vector on the right selects the second.

**Comment by fumin** [comment]
@robpike thanks for your explanation and for sharing your experience in language and systems implementation. I have few experience in language and systems implementation, so it's enlightening to learn that `panic` and `recover` can be useful in these circumstances.

I tried the expression `(5 5 ⍴ ⍳ 25)[3 2; 1 2 3]` in https://tryapl.org/ , and it returned
```
11 12 13
 6  7  8
```

I guess it makes sense for ivy to return this, too?

**Comment by kpym** [comment]
> I noticed this is still open, presumably because of the 'need an Ivy discussion forum' part since the conditional execution now exists. One possibility would be to enable GitHub Discussions on this repo. We've used them for Go discussions that are expected to be too much for the issue tracker, and the single level of threading is really helpful.
> 
> We tend to only use them for big things in Go, but projects often use them now for Q&A etc. [golang/go#47141](https://github.com/golang/go/discussions/47141) is one of our smaller discussions.

@rsc this can be moved to disucssions now ?

**Comment by robpike** [comment]
The indexing issue is fixed and https://github.com/robpike/ivy/discussions exists, so closing.
